src/compiler.ts

```diff
@@ -6512,14 +6598,64 @@ export class Compiler extends DiagnosticEmitter {
         return module.unreachable();
       }
     }
-    return this.compileCallIndirect(
-      assert(signature), // FIXME: asc can't see this yet
-      indexArg,
-      expression.arguments,
-      expression,
-      0,
-      contextualType == Type.void
-    );
+
+    var returnType = signature.returnType;
+    var tempResultLocal = this.currentFlow.getTempLocal(returnType);
+    var tempFunctionReferenceLocal = this.currentFlow.getTempLocal(returnType);
+
+    var callExpr = module.block(null, [
+      module.local_set(
+        tempFunctionReferenceLocal.index,
+        indexArg
+      ),
+      module.push(
+        module.global_get(BuiltinNames.global_closure, NativeType.I32)
+      ),
```

Comment:
Here we push the original closure context onto the stack, so that we can access it again after the call is done